File: week1/reverse-vowels-of-a-string.java
class Solution {
    public String reverseVowels(String s) {
        // Set pointers at each end of the string
        int lo = 0;
        int hi = s.length() - 1;

        // Convert string to character array
        char[] sb = s.toCharArray();
        
        while (lo < hi) {
            // Move lo and hi pointers toward center
            if (!isVowel(sb[lo])) {
                lo++;
                continue;
            }
            if (!isVowel(sb[hi])) {
                hi--;
                continue;
            }

            // When both point to vowels, switch the vowels
            char temp = sb[lo];
            sb[lo] = sb[hi];
            sb[hi] = temp;
            hi--;
            lo++;
        }
        return new String(sb);
    }
    
    // Return true if char c is a vowel
    public boolean isVowel(char c) {
        return "AEIOUaeiou".indexOf(c) != -1;
    }
}

File: spring2020/unit1/reverse-vowels-of-a-string.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        sarr = list(s)
        vowels = ["a", "e", "i", "o", "u", "A", "E", "I", "O", "U"]

        lo = 0
        hi = len(sarr) - 1
        while lo < hi:
            if sarr[lo] not in vowels:
                lo += 1
                continue

            if sarr[hi] not in vowels:
                hi -= 1
                continue

            temp = sarr[lo]
            sarr[lo] = sarr[hi]
            sarr[hi] = temp
            lo += 1
            hi -= 1

        return "".join(sarr)
